trino/auth.py

```diff
@@ -213,15 +222,17 @@ def _authenticate(self, response, **kwargs):
         response.content
         response.close()
 
-        self._thread_local.token = self._get_token(token_server, response, **kwargs)
-        return self._retry_request(response, **kwargs)
+        token = self._get_token(token_server, response, **kwargs)
+        self._store_token_to_cache(token)
```

Comment:
By default, let's not cache a token like in JDBC driver to not introduce a potential breaking change.

Reply:
Current behaviour is printing the OIDC url in the console? Could you elaborate on how the current OIDC is used in production?

I actually agree that caching the token in the keychain shouldn't be the default from a security perspective as it will persist even outside of the process context. However within the same process, given the same connection (host and username), I think it can be cached. Woud you agree?

Actually the unit test `test_multithreaded_oauth2_authentication_flow` makes you think that it only retrieves the token once per thread. However that's not actually happening because the _OAuth2TokenBearer is recreated for each http_session (every time we create a cursor a separate TrinoRequest is being created). Which makes the threadlocal storage only work per cursor. That's why i opted to make the _OAuth2TokenBearer an instance property instead of being recreated for every http_session, IMHO it should be linked to the lifetime of the connection instead of a cursor.

```

def set_http_session(self, http_session):
        http_session.auth = _OAuth2TokenBearer(http_session, self._redirect_auth_url)
        return http_session
```

Reply:
If I understand correctly, are you suggesting caching a token by default?

Reply:
The default before was already caching the token **per cursor** and per thread, which to me seems illogical. The newly default introduced in this PR is a thread-safe per connection cache, giving the user the choice to either create a connection once and share it over multiple threads or create a connection per thread. If you look at the test (test_multithreaded_oauth2_authentication_flow) you can see that before this change the token is already cached: the token is only retrieved once per thread, while there are many more requests. However the test didn't take into account that the Bearer class is instantiated for each cursor, so from a user perspective it seemed like no caching is performed.

I think for most users the keyring will provide the most seamless experience. they can either do `pip install keyring` or `pip install 'trino-python-client[local-secure-storage]'` to even get less browser popups as they will solely depend on the lifetime of the token.